### ISSUE TITLE
Locking out the qTox Client Interface window without logging off

### ISSUE THREAD

**Issue by Getron** [qTox/qTox#3314]
It would be great if we can have a "Local Pass-Code" Option to lock out the entire Tox client.

The Local Lock funktion should have options to Auto-Lock if inactive for x minutes or x hours (couple of options to selecting) into the "Privacy".

It would be very helpfull if we can add the "Two-Step verification" here to, to able to open it just with Google Authenticator or, mutch better, Open Source: FreeOTP app.

Maybe, we should add an icon to the footer black bar, a lock icon, if in options the Local Pass-Code option or Two-Step verification is activated. Icon should be here "opened lock" icon.
Maybe, each option should have own Lock icon, if activated.
And, if clicked, the locking out the Interface should be immediately, no matter how long is setted on options. On clicking the icon, icon change to "locked lock", and lock out the Interface imediately on clicking on the lock icon, and show the security code to input.

This options would be helpfull if we want to stay online, but, the Client is locked out for other people around you, like friends, co-workers or family.

**Comment by zetok** [comment]
2-step verification would be a waste of time –  #3035

**Comment by sudden6** [comment]
I don't think that should be a feature of qTox, because normally that's exactly the job of your screensaver.

**Comment by Getron** [comment]
@sudden6 you are wrong... for example, im working at my desktop, but, not alone, there is another person with me, we work together, but, i go to take an snack, and qTox is opened, or online, and the person can see my private messages, if he stay on PC to work further, until i'm back. Therefore this function is handfull: one click clossing the Interface, and open it with the code.

Ok, 2-step is maybe to much, but, this one locking is helfull.

Ok, i can go offline, or select to not showing interface on incoming messages, but, it is not handly, and take to much time to configure it, or going offline.

With the lock-system of interface, is a one click silence, but staying online.

This lock-system can be an image on top of interface, for example, and cover the interface, and would be go away just on accepted code. It should be very easy to do. This should not affect the Login Code of User. 
Unfortunatly, i can't programing it myself.

**Comment by sudden6** [comment]
Still, what you describe is exactly the job of your screensaver/screen locker. Also if you can't trust your colleague to not read your messages, but let him work at your user account, he could as well take a memory dump of the running qTox program and also read your messages.

This function you are describing only gives a false impression of security and therefore I'm strongly against implementing it.

**Comment by Getron** [comment]
![auswahl_252](https://cloud.githubusercontent.com/assets/5913218/15743360/082059e4-28c5-11e6-921e-6e07636a16cb.png)
![auswahl_253](https://cloud.githubusercontent.com/assets/5913218/15743258/71a83f40-28c4-11e6-8d68-40c52db17538.png)

@sudden6 you are totaly wrong here. Look please at Telegram, it has this function, on both, mobile and desktop. It works perfectly, looks perfectly, and is for secure reason perfect to.

Test please the Telegram Client for desktop, for example. Ok, i known, we don't want to copy them, but, if we have an idea that somebody other this implement (and Telegram is a big one Messenger), then, why we should not add to qTox to? It has to do with security, and not only security between clients, theremore security on it own machine, with a simple trick. Please, test Telegram, activate this in options of Telegram, and you will see.

It is not just when i work with friends, for example, i'm working on my PC with my customer (person), who want a web site. My qTox is online, and i have the option set that when somebody write to me, should popup the window.
Ok, my wife is writing me somthing, let say, Sexy words... ok? And my customer sitting there next to me... Ok, i can close the qTox, but, if i forggot it online? and my Wife thinking, i'm alone on PC, and therefore, for examply, she write to me words that are not for other eyes, ok?, and if window popup, my customer ... you understand the point.

Hope you understand this... devs from Telegram have this understand, and implemented it, for millions of users... why should qTox not have such a great trick to lock the client screen, but stay online, i don't understand, what is there so bad?

**Comment by sudden6** [comment]
I understand your point clearly, but I'm very much against creating the false security provided by a password to unlock qTox. Because AFAIK this can't be done in a secure way.

I think a button/shortcut to move some "curtains" over the qTox window and disable all popups would suit your described use case and make the user realize that this is only to avoid accidentally showing private stuff.

**Comment by SkyzohKey** [comment]
Or you can simply use your OS built-in lock system. Way more secure/reliable than custom code.

**Comment by zetok** [comment]
> Or you can simply use your OS built-in lock system. Way more secure/reliable than custom code.

@SkyzohKey it's not about security, but about convenience. Also, how do you use OS stuff to lock windows of a single program?

**Comment by Getron** [comment]
@sudden6 It is not a security point, it is just to hide the app interface. It is somthing like you would put an image over the interface, and move it out with an code, nothing else. Just hidding the activity of Tox App, but not close it.

Please, test Telegram, and you will see, it is a great function.

**Comment by initramfs** [comment]
@Getron Doesn't setting yourself to busy and minimzing the window achieve the same purpose? You won't be actively notified of new messages and the main qTox screen is hidden away. Better yet, minimize it to tray (if possible).

If you're worried about people snooping and actively clicking minimized windows then you have a security issue which should be solved in a secure way.

Whilst I can definitely see the usecase you propose, it does seem limited to a very narrow scenario (multiple people on same computer, you don't trust them to not open the interface but you trust them enough such that they don't steal conversations (or your decryption key for that matter) via memory dumping).

What I'm trying to say is, you seem to be requesting a solution to a problem that's orthogonal to what the solution provides. A lock screen implies a security feature, when it really isn't (or at least very hard to do right, much more so in a cross-platform way) yet what you need this lock screen for is convenience in hiding information in a non security-critical way. As such, I don't find the argument convincing enough to warrant implementation.

**Comment by Getron** [comment]
@initramfs Ok, you understand me totaly wrong here. Nobody use my computer, but, i work with my clients next to me, and we look on the same screen. Ok, maybe, i'm the only one here with this idea.

But, can you just explain us (me, and othere people here, who wotted for this idea), why, a millions people uses messenger like Telegram have implemented this, and, for millions of people is this simple idea secure, and on every devices, it work the same way, but, we (or maybe you just) don't see that is ok to have such functions?

Maybe, i forget to set the function of qTox to stay minimized if message coms, it can happen... But, this function i have proposed solved all this worryes.

And... why try then here to have a "secure" messenger, if we don't want to add such small improvments to be more, a little more secure, not just over the internet, but on private, or local, to?

Ok, we don't want to copy other messengers, i understand this, but, we have copied already almost everything from other messenger, like, for example, messaging, it self, smileys, etc...

Please, clear me this, i don't understand why is this idea bad? (don't forget, for millions of people, who uses Telegram already, everybody has this function on Telegram, on each devices, no matter, mobile, or PC's.)

Why is for Telegram, which is a very good Messenger, mature, stable and, maybe not so secure like Tox protokol, because, it use a Server to store couple of infos, but, this is not the point now, it is good idea, and for qTox not? (Don't look this just as my personaly need, i just say, that on Telegram, it works for me perfectly, and for other couple of millions people maybe to - so idea is not "just for me, or just for my uniq needs")

Thanks in advance. 
And don't get me wrong, i just don't understand the point that you say, it is not secure, and not secure, and so on... it has nothing to do with security here on qTox, security is already implemented, and is ok.

But, at the end of the day, i can't make here anything else, because, i'm not the on who implement this, or accept this, i just want to share my good experience what i have on other messengers, to have on qTox to, and, it is not a big deal to make, but, maybe, a huge help or welcome to other peoples.

Thanks again.

**Comment by initramfs** [comment]
@Getron The issue is, it's not trivial to implement a "secure" lock screen. You may think that as long as the messages are hidden and a password is required the system is secure, when in reality, this is not the case (because someone can still snoop on your messages via examining the in-use memory by qTox). But anyway, just because another client does it, it doesn't mean they are doing it correctly. There are numerous clients out there that claim security when it's doing it's doing it incorrectly or not up to the standards it claims (I mean, skype claims it's secure...) I do not know if Telegram's implementation is safe from being memory-dumped. I mean if it is, that would be something interesting to look at. But more than likely it's not.

It seems like you are satisfied with a simple lockscreen, which you want us to implement trivially and I can understand this. But you need to consider, for the more security-conscious (read: paranoid) people out there, is there a way to easily tell them that such a lock screen is inadequate to the level of security they need?

To give an analogy, imagine purchasing an expensive car with a speed dial that goes up to 300 km/h. However, unknown to you, is that the car's top speed is somewhat limited to say 240 km/h. After the purchase, you're happy with the car, because you never drive it anywhere that fast (perhaps  you top out at 140 km/h) as such the top speed limitation doesn't affect you at all.  However, other people who require speeds over 240 km/h are essentially tricked into believing the car can go that fast when in reality, it'll only top out at 240 km/h. Would it not be more accurate if the car's speed dial only went up to 240 km/h?

Having a simple lock screen implementation works in the same way, just because you are satisfied with it does not mean it is "secure" (or honest). Imagine if we added such a feature, but we implement it in an insecure way, all the people who require such a feature being implemented securely would be tricked into thinking it's secure. I'm sorry if I'm wording things poorly but client security goes quite a bit beyond just hiding messages from clients and/or requiring some sort of password. What you are asking for is extremely hard to do correctly, at least to the security standards that we require here.

**Comment by zetok** [comment]
@initramfs Um, here's the thing:

**It's not about security. Period.**

It's about simply hiding content of the window until the "unhide" button is pressed again. No password required, and it shouldn't be used anyway, since as @sudden6 pointed out (IIRC on IRC), password in qTox does guarantee security.

The question here is simple:

**Do we want to provide a button that hides content of the window until it's pressed again?**

**Comment by Getron** [comment]
@zetok That's it! Simple explanation of my proposal. That's exactly what i mean. Thanks for right explanation to other devs.

@initramfs Zetok is explained exactly what i proposed.

**Comment by Getron** [comment]
No one of devs shows little love for this feature...?

**Comment by Hasshu** [comment]
@Getron

> for example, im working at my desktop, but, not alone, there is another person with me, we work together, but, i go to take an snack, and qTox is opened, or online, and the person can see my private messages, if he stay on PC to work further, until i'm back.

If someone untrustworthy has [physical access](https://en.wikipedia.org/wiki/Physical_access) to your PC, no amount of "locking" is going to help you - be it about security or privacy.

> Ok, my wife is writing me somthing, let say, Sexy words... ok? And my customer sitting there next to me...

Frankly, I'm not sure what's more idiotic: using one's workplace PC for that kind of things, or expecting any privacy while doing it...

> Hope you understand this... devs from Telegram have this understand, and implemented it, for millions of users...

If Telegram is so awesome, why not just use it instead?

**Comment by SkyzohKey** [comment]
Telegram is a backdoor, thanks.
